javascript/ql/src/semmle/javascript/frameworks/jQuery.qll

```diff
@@ -258,144 +258,152 @@ private class JQueryChainedElement extends DOM::Element, InvokeExpr {
-            any(string method | method = "error" or method = "complete"))
+    override DataFlow::Node getAResponseDataNode(string responseType, boolean promise) {
+      (
+        responseType = getResponseType()
+        or
+        not exists(getResponseType()) and responseType = ""
+      ) and
+      promise = false and
+      (
+        result =
+          getOptionArgument([0 .. 1], "success")
               .getALocalSource()
               .(DataFlow::FunctionNode)
-              .getParameter(0))
-      or
-      result = getAnAjaxCallbackDataNode(this)
-      or
-      result = getAResponseNodeFromAnXHRObject(getAnXHRObject(this))
-    )
+              .getParameter(0)
+        or
+        result =
+          getAResponseNodeFromAnXHRObject(getOptionArgument([0 .. 1],
+              any(string method | method = "error" or method = "complete"))
+                .getALocalSource()
+                .(DataFlow::FunctionNode)
+                .getParameter(0))
+        or
+        result = getAnAjaxCallbackDataNode(this)
+        or
+        result = getAResponseNodeFromAnXHRObject(getAnXHRObject(this))
+      )
+    }
   }
-}
-
-/**
- * Gets the response data node from a call to a jqXHR Object.
- */
-DataFlow::Node getAnAjaxCallbackDataNode(ClientRequest::Range request) {
-  result =
-    request.getAMemberCall(any(string s | s = "done" or s = "then")).getCallback(0).getParameter(0)
-}
-
-/**
- * Gets the `jqXHR` object from a call to `fail` on the result from an ajax call (`request`).
- */
-DataFlow::SourceNode getAnXHRObject(ClientRequest::Range request) {
-  result = request.getAMemberCall("fail").getCallback(0).getParameter(0)
-}
 
-/**
- * Gets a node refering to the response contained in an `jqXHR` object (`obj`).
- */
-DataFlow::SourceNode getAResponseNodeFromAnXHRObject(DataFlow::SourceNode obj) {
-  result =
-    obj
-        .getAPropertyRead(any(string s |
-            s = "responseText" or
-            s = "responseXML"
-          ))
-}
+  /**
+   * Gets a response data node from a call to a method on jqXHR Object.
+   */
+  private DataFlow::Node getAnAjaxCallbackDataNode(ClientRequest::Range request) {
+    result =
+      request
+          .getAMemberCall(any(string s | s = "done" or s = "then"))
+          .getCallback(0)
+          .getParameter(0)
+  }
 
-/**
- * A model of a URL request made using a `jQuery.ajax` shorthand.
- * E.g. `jQuery.getJSON`, `jQuery.post` etc.
- * See: https://api.jquery.com/category/ajax/shorthand-methods/.
- *
- * The method signatures:
- * jQuery.get( url [, data ] [, success ] [, dataType ] )
- * jQuery.getJSON( url [, data ] [, success ] )
- * jQuery.getScript( url [, success ] )
- * jQuery.post( url [, data ] [, success ] [, dataType ] )
- * .load( url [, data ] [, complete ] )
- */
-private class JQueryAjaxShortHand extends ClientRequest::Range {
-  string name;
+  /**
+   * Gets a `jqXHR` object from a call to the `fail(..)` method on jqXHR Object.
```

Comment:
This docstring should be near identical to the one for  `getAnAjaxCallbackDataNode` , shouldn't it?

Reply:
Not necessarily, this docstring describes getting an `jqXHR` object, and not a response node.  
I think I'll change the predicate such that it directly computes a response node.